File: gui/app/lib/skills.ts
/**
 * 「skills」画面（ADR-0056 D3 続き、docs/celeris-api-v1.md §3.112〜3.117。Phase 82 / G35）の
 * **純粋な**表示ロジック。
 *
 * `~/lib/knowledge.ts` と同じ流儀: celeris が返すもの（`root` / `mounted_by` / 検証結果）は
 * 一切作り直さない。ここでやるのは「送る前に明らかに通らない frontmatter を止める」「雛形を出す」
 * 「継いだ mount と自分の mount を分ける（GUI は継承を再計算しないが、`own` と `inherited` に分けて
 * 見せるのは `node.profile.skills_mounts` と `effective.skills_mounts` の**差分を取るだけ**なので
 * 再計算ではない）」だけ。判定（frontmatter の検証・mount の可否）は celeris 側にある。
 */

/** skill 名の綴り（`task_core::knowledge::is_valid_skill_name` と同じ規則）。 */
export function isValidSkillName(name: string): boolean {
  return name.length > 0 && name.length <= 64 && /^[a-z0-9-]+$/.test(name);
}

/** 「skills」画面の URL（選んでいる skill・編集モードを保つ）。 */
export function skillsHref(options: { name?: string | null; edit?: boolean; create?: boolean } = {}): string {
  const params = new URLSearchParams();
  if (options.name) params.set("name", options.name);
  if (options.edit) params.set("edit", "1");
  if (options.create) params.set("create", "1");
  const query = params.toString();
  return `/knowledge/skills${query ? `?${query}` : ""}`;
}

/** 新しい skill の `SKILL.md` の雛形（Claude Code の skills 形式。frontmatter の `name`/`description` 必須）。 */
export function skillMarkdownTemplate(name: string): string {
  const skillName = name || "skill-name";
  return `---\nname: ${skillName}\ndescription: この skill が何をするかを 1〜2 文で\n---\n\n# ${skillName}\n\n手順や参照するべき情報をここに書きます。\n`;
}

/** `SKILL.md` の frontmatter から読み取れるもの（celeris が持つ `task_ops::knowledge::skill_frontmatter` と
 * 同じ「`---\nkey: value\n---`だけを読む最小パーサ」の TypeScript 版。プレビュー用で、正本は celeris が
 * `PUT /skills/{name}` で行う検証）。 */
export interface SkillFrontMatter {
  name: string | null;
  description: string | null;
  source: string | null;
}

export function parseSkillFrontMatter(raw: string): SkillFrontMatter {
  const out: SkillFrontMatter = { name: null, description: null, source: null };
  const text = raw.startsWith("﻿") ? raw.slice(1) : raw;
  if (!text.startsWith("---")) return out;
  const afterOpen = text.slice(3).replace(/^\n/, "");
  const closeAt = afterOpen.indexOf("\n---");
  if (closeAt < 0) return out;
  const body = afterOpen.slice(0, closeAt);
  for (const line of body.split("\n")) {
    const cut = line.indexOf(":");
    if (cut < 0) continue;
    const key = line.slice(0, cut).trim();
    const value = line.slice(cut + 1).trim();
    if (key === "name") out.name = value;
    else if (key === "description") out.description = value;
    else if (key === "source") out.source = value;
  }
  return out;
}

/** front matter を切り離す（先頭の `---` … `---`）。`~/lib/knowledge.ts::splitFrontMatter` と同じ考え方。 */
export function splitSkillFrontMatter(raw: string): { frontMatter: string; body: string } {
  const text = raw.startsWith("﻿") ? raw.slice(1) : raw;
  if (!text.startsWith("---")) return { frontMatter: "", body: text };
  const afterOpen = text.slice(3).replace(/^\n/, "");
  const closeAt = afterOpen.indexOf("\n---");
  if (closeAt < 0) return { frontMatter: "", body: text };
  const rest = afterOpen.slice(closeAt + 4);
  return { frontMatter: afterOpen.slice(0, closeAt), body: rest.startsWith("\n") ? rest.slice(1) : rest };
}

/**
 * SKILL.md を描く前の書き換え: front matter を落とし（構造では別に見せるため）、見出しを 1 段落とす
 * （`# <name>` から始まる本文が多く、カード自身の見出し（`<h2>`）と並ぶと「1 画面に h1 は 1 つ」
 * （ADR-0055 D1）が崩れるため。`~/lib/docs.ts::prepareDocBody` と同じ「表示のための書き換え」の考え方）。
 */
export function skillMarkdownBody(raw: string): string {
  const { body } = splitSkillFrontMatter(raw);
  return body
    .split("\n")
    .map((line) => (/^#{1,5} /.test(line) ? `#${line}` : line))
    .join("\n");
}

/**
 * 送る前の frontmatter の検査（celeris の 422 `validation` を先に人に見せるだけ。
 * **判定の正本は celeris**）。通るなら `null`、通らないなら理由。
 */
export function skillMarkdownProblem(name: string, skillMd: string): string | null {
  if (!isValidSkillName(name)) return "skill 名は英小文字・数字・ハイフンだけ（1〜64 文字）にしてください。";
  const front = parseSkillFrontMatter(skillMd);
  if (front.name === null) return "先頭に `---` で始まる frontmatter が要ります（`name` / `description`）。";
  if (front.name !== name) return `frontmatter の \`name: ${front.name ?? ""}\` が URL の skill 名と違います。`;
  if (!front.description || front.description.trim() === "") return "frontmatter の `description` を書いてください。";
  return null;
}

/**
 * ノードの mount を「自分で mount したもの」と「継いだもの（親から）」に分ける。
 * `own` は `node.profile.skills_mounts`、`effective` は `EffectiveProfile.skills_mounts`
 * （継いだ後、`GET /org` が返す）。`inherited` は `effective` にあって `own` に無いもの
 * （GUI は継承の計算をやり直さない。差分を取るだけ）。
 */
export interface MountedSkills {
  /** そのノード自身が mount しているもの（そこで unmount できる）。 */
  own: string[];
  /** 親から継いだもの（そのノードでは unmount できない。親で外す）。 */
  inherited: string[];
}

export function splitMountedSkills(own: readonly string[] = [], effective: readonly string[] = []): MountedSkills {
  const ownSet = new Set(own);
  return {
    own: [...own],
    inherited: effective.filter((skill) => !ownSet.has(skill)),
  };
}

File: crates/task-api/src/skills.rs
//! Phase 82（ADR-0056 D3 続き）: skills を GUI から見る・作る・mount する。
//!
//! - `GET  /skills` — KB の `skills/` の一覧（name / description / updated / mounted_by）。読み取り
//! - `GET  /skills/{name}` — `SKILL.md` 本文と付属ファイルの一覧。読み取り
//! - `PUT  /skills/{name}` — **管理系**。`task_ops::knowledge::skills_put` と同じ検証（frontmatter の
//!   `name` / `description` 必須）。celeris-mcp の `skills_put` ツールと**同じ関数**を呼ぶ
//! - `DELETE /skills/{name}` — **管理系**。どこかの組織ノードに mount されていれば 409 `skill_mounted`
//! - `POST /org/{id}/skills` `{skill}` / `DELETE /org/{id}/skills/{skill}` — **管理系**。mount / unmount。
//!   celeris-mcp の `org_mount_skill`/`org_unmount_skill` と**同じ** `task_ops::knowledge::set_skill_mount`
//!   を呼ぶので挙動は同一（ADR-0056 D2/D3）
//!
//! `mounted_by` は「継いだ後」（`EffectiveProfile.skills_mounts`）で判定する: 親ノードで mount すれば
//! 子ノードにもそこに現れる。組織の木を GUI が再計算しない規律（ADR-0046 D1）に合わせ、平坦化は
//! `task_core::resolve_profile` に任せる。

use std::collections::HashMap;

use axum::body::Body;
use axum::http::{HeaderMap, StatusCode};
use axum::response::IntoResponse;
use axum::routing::{get, post};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use task_core::{OrgNode, TaskStore};
use task_ops::knowledge::{self as ops_kb, SkillError};
use time::OffsetDateTime;

use crate::handlers::{ApiResult, Params, json_response, load_org_node, no_query, read_json};
use crate::knowledge::{require_kb, root_of};
use crate::middleware::require_admin;
use crate::problem::{ApiProblem, store_problem};
use crate::state::ApiState;
use crate::types::ValidationError;

pub(crate) fn routes() -> axum::Router<ApiState> {
    axum::Router::new()
        .route("/api/v1/skills", get(list_skills))
        .route(
            "/api/v1/skills/{name}",
            get(get_skill).put(put_skill).delete(delete_skill),
        )
        .route("/api/v1/org/{id}/skills", post(mount_skill))
        .route(
            "/api/v1/org/{id}/skills/{skill}",
            axum::routing::delete(unmount_skill),
        )
}

// ---------------------------------------------------------------------------
// 応答の型（`docs/gui/api.md` §3.112〜）
// ---------------------------------------------------------------------------

/// `GET /skills` の 1 件。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct SkillSummaryView {
    pub name: String,
    pub description: String,
    /// 最後のコミットの時刻（RFC 3339。無ければ省略）。
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated: Option<String>,
    /// この skill を（継承も含め）mount している組織ノードの id。
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub mounted_by: Vec<String>,
}

/// `GET /skills`。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct SkillList {
    /// KB の根（絶対パス）。
    pub root: String,
    /// `celerisctl knowledge init` が済んでいるか。偽なら `items` は空。
    pub initialized: bool,
    pub items: Vec<SkillSummaryView>,
}

/// `GET /skills/{name}`。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct SkillDetailView {
    pub name: String,
    /// `SKILL.md` の中身（frontmatter を含む）。
    pub skill_md: String,
    /// 同じディレクトリの付属ファイル（相対パス。`SKILL.md` 自身は含まない）。
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub files: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub mounted_by: Vec<String>,
}

/// `PUT /skills/{name}` の付属ファイル 1 件。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct SkillFileBody {
    pub path: String,
    pub content: String,
}

/// `PUT /skills/{name}` の本文。
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct SkillPutBody {
    /// `SKILL.md` の中身（frontmatter を含む。`name` / `description` 必須、`name` はこの URL の
    /// `{name}` と一致していること）。
    pub skill_md: String,
    #[serde(default)]
    pub files: Vec<SkillFileBody>,
}

/// `PUT /skills/{name}` の応答。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct SkillPutResult {
    /// `skills/<name>/SKILL.md`。
    pub path: String,
}

/// `POST /org/{id}/skills` の本文。
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct OrgSkillMountBody {
    pub skill: String,
}

// ---------------------------------------------------------------------------
// 共通
// ---------------------------------------------------------------------------

fn skill_not_found(name: &str) -> ApiProblem {
    ApiProblem::new(
        StatusCode::NOT_FOUND,
        "skill_not_found",
        format!("skill not found: {name}"),
    )
}

fn skill_mounted(name: &str, nodes: &[String]) -> ApiProblem {
    ApiProblem::new(
        StatusCode::CONFLICT,
        "skill_mounted",
        format!(
            "skill {name:?} is mounted by: {}（先に外してから消す）",
            nodes.join(", ")
        ),
    )
}

/// [`SkillError`] → [`ApiProblem`]。`Failed`（I/O 失敗）だけ 500、それ以外は入力の誤りなので 422。
fn skill_error_to_problem(e: SkillError) -> ApiProblem {
    match e {
        SkillError::Failed(detail) => ApiProblem::internal(detail),
        other => ApiProblem::validation(vec![ValidationError {
            field: None,
            message: other.to_string(),
        }]),
    }
}

/// その skill を（継承も含め）mount している組織ノードの id を、全ノード分まとめて 1 度に計算する
/// （`GET /skills` が N 件でも `org_list` は 1 回だけ）。
fn mounted_by_map(nodes: &[OrgNode]) -> HashMap<String, Vec<String>> {
    let mut map: HashMap<String, Vec<String>> = HashMap::new();
    for node in nodes {
        let effective = task_core::resolve_profile(nodes, &node.id);
        for skill in effective.skills_mounts {
            map.entry(skill).or_default().push(node.id.clone());
        }
    }
    map
}

fn skill_updated(root: &std::path::Path, name: &str) -> Option<String> {
    ops_kb::history(root, &format!("skills/{name}/SKILL.md"))
        .into_iter()
        .next()
        .map(|c| c.at)
}

// ---------------------------------------------------------------------------
// GET /skills
// ---------------------------------------------------------------------------

async fn list_skills(
    axum::extract::State(state): axum::extract::State<ApiState>,
    axum::extract::RawQuery(raw): axum::extract::RawQuery,
) -> ApiResult {
    no_query(&raw)?;
    let root = root_of(&state)?;
    let view = state
        .blocking(move |store| {
            let initialized = ops_kb::exists(&root);
            if !initialized {
                return Ok(SkillList {
                    root: root.display().to_string(),
                    initialized,
                    items: Vec::new(),
                });
            }
            let nodes = store.org_list().map_err(store_problem)?;
            let by = mounted_by_map(&nodes);
            let items: Vec<SkillSummaryView> = ops_kb::skills_list(&root)
                .into_iter()
                .map(|s| SkillSummaryView {
                    updated: skill_updated(&root, &s.name),
                    mounted_by: by.get(&s.name).cloned().unwrap_or_default(),
                    name: s.name,
                    description: s.description,
                })
                .collect();
            Ok(SkillList {
                root: root.display().to_string(),
                initialized,
                items,
            })
        })
        .await?;
    Ok(json_response(StatusCode::OK, &view))
}

// ---------------------------------------------------------------------------
// GET /skills/{name}
// ---------------------------------------------------------------------------

async fn get_skill(
    axum::extract::State(state): axum::extract::State<ApiState>,
    Params(name): Params<String>,
    axum::extract::RawQuery(raw): axum::extract::RawQuery,
) -> ApiResult {
    no_query(&raw)?;
    let root = root_of(&state)?;
    let view = state
        .blocking(move |store| {
            let detail = ops_kb::skills_get(&root, &name).ok_or_else(|| skill_not_found(&name))?;
            let nodes = store.org_list().map_err(store_problem)?;
            let mounted_by = mounted_by_map(&nodes).remove(&name).unwrap_or_default();
            let updated = skill_updated(&root, &name);
            Ok(SkillDetailView {
                name: detail.name,
                skill_md: detail.skill_md,
                files: detail.files,
                updated,
                mounted_by,
            })
        })
        .await?;
    Ok(json_response(StatusCode::OK, &view))
}

// ---------------------------------------------------------------------------
// PUT /skills/{name}（管理系）
// ---------------------------------------------------------------------------

async fn put_skill(
    axum::extract::State(state): axum::extract::State<ApiState>,
    headers: HeaderMap,
    Params(name): Params<String>,
    axum::extract::RawQuery(raw): axum::extract::RawQuery,
    body: Body,
) -> ApiResult {
    no_query(&raw)?;
    require_admin(&state, &headers)?;
    let request: SkillPutBody = read_json(body, false).await?;
    let root = root_of(&state)?;
    let result = state
        .blocking(move |_| {
            require_kb(&root)?;
            let files: Vec<(String, String)> = request
                .files
                .into_iter()
                .map(|f| (f.path, f.content))
                .collect();
            // GUI からの書き込みは人の操作だが、`source` は「どこから来たか」を frontmatter に残す
            // ADR-0056 D3 の趣旨（MCP は `mcp:<client_id>`）に合わせ、GUI は `gui` を残す（冪等: 既に
            // `source:` があれば `skills_put` が触らない）。
            match ops_kb::skills_put(&root, &name, &request.skill_md, &files, Some("gui")) {
                Ok(path) => {
                    tracing::info!(who = "admin", op = "skill_put", name = %name, "admin: skill written");
                    Ok(SkillPutResult { path })
                }
                Err(e) => Err(skill_error_to_problem(e)),
            }
        })
        .await?;
    Ok(json_response(StatusCode::OK, &result))
}

// ---------------------------------------------------------------------------
// DELETE /skills/{name}（管理系）
// ---------------------------------------------------------------------------

async fn delete_skill(
    axum::extract::State(state): axum::extract::State<ApiState>,
    headers: HeaderMap,
    Params(name): Params<String>,
    axum::extract::RawQuery(raw): axum::extract::RawQuery,
) -> ApiResult {
    no_query(&raw)?;
    require_admin(&state, &headers)?;
    let root = root_of(&state)?;
    state
        .blocking(move |store| {
            require_kb(&root)?;
            if ops_kb::skills_get(&root, &name).is_none() {
                return Err(skill_not_found(&name));
            }
            let nodes = store.org_list().map_err(store_problem)?;
            let mounted_by = mounted_by_map(&nodes).remove(&name).unwrap_or_default();
            if !mounted_by.is_empty() {
                return Err(skill_mounted(&name, &mounted_by));
            }
            match ops_kb::skills_delete(&root, &name) {
                Ok(_) => {
                    tracing::info!(who = "admin", op = "skill_delete", name = %name, "admin: skill deleted");
                    Ok(())
                }
                Err(e) => Err(skill_error_to_problem(e)),
            }
        })
        .await?;
    Ok(StatusCode::NO_CONTENT.into_response())
}

// ---------------------------------------------------------------------------
// POST /org/{id}/skills、DELETE /org/{id}/skills/{skill}（管理系）
// ---------------------------------------------------------------------------

async fn mount_skill(
    axum::extract::State(state): axum::extract::State<ApiState>,
    headers: HeaderMap,
    Params(id): Params<String>,
    axum::extract::RawQuery(raw): axum::extract::RawQuery,
    body: Body,
) -> ApiResult {
    no_query(&raw)?;
    require_admin(&state, &headers)?;
    let request: OrgSkillMountBody = read_json(body, false).await?;
    let skill = request.skill;
    let skill_for_log = skill.clone();
    let node = state
        .blocking(move |store| -> Result<OrgNode, ApiProblem> {
            let mut node = load_org_node(store, &id)?;
            ops_kb::set_skill_mount(&mut node.profile.skills_mounts, &skill, true)
                .map_err(skill_error_to_problem)?;
            node.updated_at = OffsetDateTime::now_utc();
            store.org_upsert(&node).map_err(store_problem)
        })
        .await?;
    tracing::info!(who = "admin", op = "org_skill_mount", org_id = %node.id, skill = %skill_for_log, "admin: skill mounted");
    Ok(json_response(StatusCode::OK, &node))
}

async fn unmount_skill(
    axum::extract::State(state): axum::extract::State<ApiState>,
    headers: HeaderMap,
    Params((id, skill)): Params<(String, String)>,
    axum::extract::RawQuery(raw): axum::extract::RawQuery,
) -> ApiResult {
    no_query(&raw)?;
    require_admin(&state, &headers)?;
    let node = state
        .blocking(move |store| -> Result<OrgNode, ApiProblem> {
            let mut node = load_org_node(store, &id)?;
            ops_kb::set_skill_mount(&mut node.profile.skills_mounts, &skill, false)
                .map_err(skill_error_to_problem)?;
            node.updated_at = OffsetDateTime::now_utc();
            store.org_upsert(&node).map_err(store_problem)
        })
        .await?;
    tracing::info!(who = "admin", op = "org_skill_unmount", org_id = %node.id, "admin: skill unmounted");
    Ok(json_response(StatusCode::OK, &node))
}
